### DISCUSSION TITLE
Support concurrent BackgroundTasks

### DISCUSSION THREAD

**Discussion by mbforbes** [fastapi/fastapi#10682] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import asyncio
from fastapi import BackgroundTasks

class ConcurrentBackgroundTasks(BackgroundTasks):
    """This is all it takes to implement"""
    async def __call__(self) -> None:
        await asyncio.gather(*[task() for task in self.tasks])

# This doesn't work; still constructed as BackgroundTasks
@app.post("/foo")
def foo(background_tasks: ConcurrentBackgroundTasks):
    pass

# Works, but we must now manually add back FastAPI schema and validation features
@app.post("/foo", response_model=Model)
def foo():
    bgt = ConcurrentBackgroundTasks()
    bgt.add_task(...)
    return JSONResponse(
        content=jsonable_encoder(
            Model(...)
        ),
        background=bgt,
    )
```

### Description

**tl;dr: I need to run background tasks concurrently. I found a way, but it required me to spend many hours figuring things out. It's essentially a difficult one-line change that makes FastAPI messier. It'd be delightful if FastAPI supported this directly.**

I need to run background tasks concurrently.

My use case: make multiple HTTP requests after a response in a realtime application. As each request takes ~2s, if I can run them concurrently, N tasks complete in 2s rather than 2Ns.

I'd like to propose a feature request to either 
1. Allow concurrently-executing background tasks
2. Allow providing a custom BackgroundTask subclass in the path operation function

My implementation above works, but it requires returning a raw starlette Response. This means we have to manually add back schema generation (`response_model`), validation (`Model(...)`) and serialization (`jsonable_encoder(...)`). I think it'd be more elegant if FastAPI supported idea (1) or (2).

Here's a sketch of each idea.

## 1. Allow concurrently-executing background tasks

In the [BackgroundTasks implementation](https://github.com/tiangolo/fastapi/blob/master/fastapi/background.py), provide an alternate instance or `__call__` implementation to allow parallel execution. (The implementation is the one line I posted above, directly from https://github.com/encode/starlette/issues/549#issue-456358615.) I'm a novice to FastAPI internals so I'm not sure exactly how concurrent vs sequential would specified in, say, a path operation function.

I opened a discussion on starlette (https://github.com/encode/starlette/discussions/2338) to see whether they'd consider supporting a `concurrent` flag. The idea has been previously mentioned there (https://github.com/encode/starlette/issues/549).

## 2. Allow providing a custom BackgroundTask subclass in the path operation function

I.e., make the following work:

```py
# Proposed: respect ConcurrentBackgroundTasks here rather than just returning a vanilla BackgroundTasks
@app.post("/foo")
def foo(background_tasks: ConcurrentBackgroundTasks):
    pass
```

This was asked previously in #8478, with the answer that it is impossible. A great comment there (https://github.com/tiangolo/fastapi/discussions/8478#discussioncomment-5152355) dug through the FastAPI internals to see why this can't be overridden.

Detection:

https://github.com/tiangolo/fastapi/blob/ac93277d3b506f4076aee1af6ae5a86406f545c6/fastapi/dependencies/utils.py#L309-L311

Setting (always constructs `BackgroundTasks` instance).

https://github.com/tiangolo/fastapi/blob/ac93277d3b506f4076aee1af6ae5a86406f545c6/fastapi/dependencies/utils.py#L631-L634

Would it be possible to instantiate the provided subclass instead?

---

Many thanks for your consideration, and apologies if this isn't the right place for a feature request. FastAPI has been delightful and I am grateful for it.

### Operating System

macOS

### FastAPI Version

0.104.1

### Pydantic Version

2.5.0

### Python Version

3.10.6

**Comment by YuriiMotov** [comment, marked as answer]
Interesting idea, but I'm afraid I think it will not be implemented.

Here is why:

<details>

It's not as simple as it seems to implement this.
Imagine you added `bg_tasks: ConcurrentBackgroundTasks` as a parameter to one dependency and `bg_tasks: BackgroundTasks` to another dependency that are both used in the same endpoint. How should FastAPI resolve this? In current implementation both `bg_tasks` will point to the same instance of `BackgroundTasks`.

Alternative approach could be to add a parameter like `bg_tasks_class` to `FastAPI()` and then use this class to create an instance:

```python
# This doesn't work, it's just to illustrate possible approach

app = FastAPI(bg_tasks_class=ConcurrentBackgroundTasks)

@app.post("/foo")
def foo(response: Response, background_tasks: BackgroundTasks):
    assert isinstance(background_tasks, ConcurrentBackgroundTasks)
```


But.. As it's been said a lot of times, FastAPI tends not to add features if it's possible and easy to implement this without adding this feature.

</details>

You can just combine your tasks in one function and add it as a task:

```python
import asyncio
from fastapi import BackgroundTasks, FastAPI

async def concurrent_tasks(tasks: list[asyncio.Task]) -> None:
    await asyncio.gather(*tasks)

app = FastAPI()


async def sleeeep(n: int):
    print(f"{n}: Sleeping...")
    await asyncio.sleep(2)
    print(f"{n}: Done sleeping!")

@app.post("/foo")
def foo(background_tasks: BackgroundTasks):
    tasks = [sleeeep(n) for n in range(1, 4)]
    background_tasks.add_task(concurrent_tasks, tasks)    
    return {"message": "foo"}

```

**Reply by mbforbes** [reply]
Thanks for the reply!

Do you think it'd be worth an addition to the docs?

Someone more experienced may find it simple and easy to do, but it took me a while to figure out the "correct" way to do this (and convince myself it wasn't supported as a first-class concept).

**Reply by afzalIbnSH-locus** [reply]
Sharing @mbforbes 's sentiment, adding this to the doc would be really helpful. Like [how starlette has](https://www.starlette.io/background/) in the BackgroundTasks section in an "Important" callout box at the bottom.

**Reply by ScrimForever** [reply]
Is possible do it using new uvloop concorrence , so for any backgroundtasks it will create another "process id". 
I think is good if its can be default or bool parameter "concorrence" on start of app
